drivers: sensor: max30208: Added driver and sample for MAX30208

trembel opened the pull request:
Added driver and sample for MAX30208 Maxim Integrated I2C temperature
sensor. Tested on nrf52dk.

Signed-off-by: Silvano Cortesi <silvano.cortesi@hotmail.com>

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,234 @@
+}
+
+static struct max30208_config max30208_config = {
+	.i2c_label = DT_INST_BUS_LABEL(0),
+	.i2c_addr = DT_INST_REG_ADDR(0),
```
better use i2c_dt_spec

trembel replied:
I see. Dev began before the addition of `i2c_dt_spec`. How can I rebase my branch without losing my changes to a newer version?

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,234 @@
+	}
+
+	/* Check the part id to make sure this is MAX30208 */
+	if (i2c_reg_read_byte(data->i2c, config->i2c_addr, MAX30208_REG_PART_ID, &part_id)) {
+		LOG_ERR("Could not get Part ID");
+		return -EIO;
+	}
+	if (part_id != MAX30208_PART_ID) {
+		LOG_ERR("Got Part ID 0x%02x, expected 0x%02x", part_id, MAX30208_PART_ID);
+		return -EIO;
+	}
+
+	/* Reset the sensor */
+	if (i2c_reg_write_byte(data->i2c, config->i2c_addr, MAX30208_REG_SYS_CTRL,
+			       MAX30208_RESET_MASK)) {
+		LOG_ERR("Failed to reset sensor!");
+		return -EIO;
+	}
+	/* Wait for reset to be cleared */
+	k_sleep(K_MSEC(10));
+	do {
+		if (i2c_reg_read_byte(data->i2c, config->i2c_addr, MAX30208_REG_SYS_CTRL,
+				      &sys_ctrl)) {
+			LOG_ERR("Couldn't read system control after reset");
+			return -EIO;
+		}
+	} while (sys_ctrl & MAX30208_RESET_MASK);
+
+	/* Write the FIFO configuration register */
+	if (i2c_reg_write_byte(data->i2c, config->i2c_addr, MAX30208_REG_FIFO_CFG1,
+			       config->fifo.fifo_a_full)) {
+		LOG_ERR("Failed to initialize FIFO config 1!");
+		return -EIO;
+	}
+	if (i2c_reg_write_byte(data->i2c, config->i2c_addr, MAX30208_REG_FIFO_CFG2,
+			       MAX30208_STC_TO_B(config->fifo.config2))) {
+		LOG_ERR("Failed to initialize FIFO config 2!");
+		return -EIO;
+	}
```
error codes need to be propagated,

```c
ret = f();
if (ret < 0) {
    return ret;
}
```

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,234 @@
+		return -EIO;
+	}
+	/* Wait for reset to be cleared */
+	k_sleep(K_MSEC(10));
```
constant needs to be defined

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,234 @@
+	return 0;
+}
+
+static int max30208_sample_fetch(const struct device *dev, enum sensor_channel chan)
```
requested channel is not checked here (all or ambient temp)

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,234 @@
+	}
+
+	/* Concat two bytes to one sample */
+	data->raw = (int16_t)(local_buffer[0] << 8) | local_buffer[1];
```
please, use sys/byteorder.h API

gmarull commented on drivers/sensor/max30208/Kconfig:
```diff
@@ -0,0 +1,103 @@
+	help
+	  Controls the behavior of the FIFO when the FIFO becomes completely
+	  filled with data. If set, the FIFO address rolls over to zero and the
+	  FIFO continues to fill with new data. If not set, then the FIFO is
+	  not updated until FIFO_DATA is read or the WRITE/READ pointer
+	  positions are changed.
+
+config MAX30208_FIFO_A_FULL_TYPE
+	bool "FIFO almost full type"
+	help
+	  If false, then the A_FULL status bit is asserted on each sample for
+	  which the condition persists. If true, then the bit is asserted only
+	  once until the next A_FULL condition.
+
+config MAX30208_FIFO_STAT_CLR
+	bool "FIFO status clear"
+	help
+	  If true, then A_FULL and TEMP_RDY bit are cleared by a FIFO_DATA register read.
+
+config MAX30208_GPIO0_MODE
+	int "GPIO 0 mode"
+	range 0 3
+	default 2
+	help
+	    Set the mode for GPIO0.
+	    0 = High impedance input
+	    1 = Output (open-drain)
+	    2 = Input with 1MOhm pulldown
+	    3 = Interrupt output (active low)
+
+config MAX30208_GPIO1_MODE
+	int "GPIO 1 mode"
+	range 0 3
+	default 2
+	help
+	    Set the mode for GPIO1.
+	    0 = High impedance input
+	    1 = Output (open-drain)
+	    2 = Input with 1MOhm pulldown
+	    3 = Convert Temperature input (active low)
```
these options can likely be moved to devicetree, so that they can be configured per-instance.

trembel replied:
can they be moved to dev tree but still have a Kconfig being dependend on them (interrupt requires GPIO0 mode being 3)?

gmarull commented on drivers/sensor/max30208/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+zephyr_library()
+
+zephyr_library_sources(max30208.c)
+zephyr_library_sources_ifdef(CONFIG_MAX30208_TRIGGER    max30208_trigger.c)
```
```suggestion
zephyr_library_sources_ifdef(CONFIG_MAX30208_TRIGGER max30208_trigger.c)
```

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,234 @@
+
+static struct max30208_data max30208_data;
+
+DEVICE_DT_INST_DEFINE(0, max30208_init, NULL, &max30208_data, &max30208_config, POST_KERNEL,
```
please, make driver multi-instance (see most other sensors)

gmarull commented on drivers/sensor/max30208/max30208.h:
```diff
@@ -0,0 +1,219 @@
+
+/* register structs */
+#ifdef CONFIG_BIG_ENDIAN
+struct MAX30208_status {
```
naming convention:
```suggestion
struct max30208_status {
```

trembel replied:
what for enums? also lower case for its name (and then inside upper case)?

gmarull commented on drivers/sensor/max30208/max30208_trigger.c:
```diff
@@ -0,0 +1,240 @@
+		gpio_add_callback(data->gpio, &data->gpio_cb);
+	}
+
+	if (rc != 0) {
```
this can be moved above, so `rc == 0` branch is not required

gmarull commented on drivers/sensor/max30208/max30208_trigger.c:
```diff
@@ -0,0 +1,240 @@
+	int rc = gpio_pin_configure(data->gpio, config->int_pin, GPIO_INPUT | config->int_flags);
+
+	if (rc == 0) {
+		gpio_add_callback(data->gpio, &data->gpio_cb);
```
error check missing here

gmarull commented on drivers/sensor/max30208/max30208_trigger.c:
```diff
@@ -0,0 +1,240 @@
+
+	data->status = 0;
+
+	data->gpio = device_get_binding(config->int_name);
```
please, move to `gpio_dt_spec`

gmarull commented on samples/sensor/max30208/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
sample should go into its own separate commit

gmarull reviewed (changes requested):
Thanks for this addition, left a few comments

trembel committed:
drivers: sensor: max30208: Added driver for MAX30208

Added driver for MAX30208 Maxim Integrated I2C temperature
sensor. Tested on nrf52dk.

Signed-off-by: Silvano Cortesi <silvano.cortesi@hotmail.com>

gmarull commented on drivers/sensor/max30208/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+# Makefile - MAX30208 digital temperature sensor
+#
```
nit: not a Makefile :-)

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,346 @@
+static int max30208_channel_get(const struct device *dev, enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	if (chan != SENSOR_CHAN_AMBIENT_TEMP) {
```
`SENSOR_CHAN_ALL` should also be supported, see https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/sensor.h#L513

trembel replied:
but this is channel get, shouldn't it be specified here?

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,346 @@
+	if (chan != SENSOR_CHAN_AMBIENT_TEMP) {
+		return -ENOTSUP;
+	}
+	struct max30208_data *data = dev->data;
+	uint8_t local_buffer[MAX30208_BYTES_PER_VAL] = { 0 };
```
better move declarations/init to the top.

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,346 @@
+	struct max30208_status status;
+
+	/* Read data ready */
+	int ret =
+		i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_INT_STS, MAX30208_STC_TO_P(status));
```
doesn't seem to follow styleguide:
```
int ret = f(a, very, long,
            list);
```

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,346 @@
+		i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_INT_STS, MAX30208_STC_TO_P(status));
+
+	if (ret < 0) {
+		LOG_ERR("Timeout while waiting on MAX30208 data ready\n");
```
`ret < 0` from `i2c_reg_read_byte_dt` doesn't imply a timeout, afaik.

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,346 @@
+
+	uint8_t fifo_data_counter = 0;
+	int ret =
+		i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_FIFO_DATA_CTR, &fifo_data_counter);
+
+	if (ret < 0) {
+		LOG_ERR("Timeout while waiting on FIFO data counter\n");
```
ditto

gmarull commented on drivers/sensor/max30208/max30208.h:
```diff
@@ -0,0 +1,220 @@
+/* enums and definitions */
+#define MAX30208_RESET_MASK 0x01
+#define MAX30208_CONVERT_T_MASK 0xFF
+#define MAX30208_INT_A_FULL_MASK (1 << 7)
+#define MAX30208_INT_TEMP_LO_MASK (1 << 2)
+#define MAX30208_INT_TEMP_HI_MASK (1 << 1)
+#define MAX30208_INT_TEMP_RDY_MASK (1 << 0)
```
BIT() macro should be used

gmarull commented on drivers/sensor/max30208/max30208.h:
```diff
@@ -0,0 +1,220 @@
+#define MAX30208_INT_TEMP_LO_MASK (1 << 2)
+#define MAX30208_INT_TEMP_HI_MASK (1 << 1)
+#define MAX30208_INT_TEMP_RDY_MASK (1 << 0)
+#define MAX30208_GPIO_MODE_MASK(gpioX, mode) (mode << (gpioX * 6))
```
`6U`

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,346 @@
+static int max30208_data_ready(const struct device *dev, uint8_t *ready)
+{
+	const struct max30208_config *config = dev->config;
+	struct max30208_status status;
```
why using a struct to read a single byte?

trembel replied:
Its a bitfield struct. As it implies the same runtime cost as using a raw byte, I think its better here as specific fields can be accessed by a meaningful name.

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,346 @@
+	const struct max30208_config *config = dev->config;
+	struct max30208_data *data = dev->data;
+
+	uint8_t fifo_data_counter = 0;
```
`0U`

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,346 @@
+		if (ring_buf_space_get(&data->raw_buffer) < MAX30208_BYTES_PER_VAL) {
+			struct sensor_value lost_val;
+
+			ret = max30208_channel_get(dev, SENSOR_CHAN_AMBIENT_TEMP, &lost_val);
+			if (ret < 0) {
+				return ret;
+			}
+			LOG_INF("Buffer size to small. Value %d.%d is lost", lost_val.val1,
+				lost_val.val2);
```
is it worth reading just to log? what if logging is disabled? Isn't it sufficient to warn about buffer being too small?

trembel replied:
since one item must be poped on a full ringbuffer, I could just pop and not store it, but I don't see a lot of advantage here..

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,346 @@
+	return 0;
+}
+
+#ifdef CONFIG_MAX30208_FIFO_ROLLOVER_EN
+#define MAX30208_FIFO_RO .fifo_ro = CONFIG_MAX30208_FIFO_ROLLOVER_EN,
+#else
+#define MAX30208_FIFO_RO .fifo_ro = 0,
+#endif
+
+#ifdef CONFIG_MAX30208_FIFO_A_FULL_TYPE
+#define MAX30208_A_FULL_TYPE .a_full_type = CONFIG_MAX30208_FIFO_A_FULL_TYPE,
+#else
+#define MAX30208_A_FULL_TYPE .a_full_type = 0,
+#endif
+
+#ifdef CONFIG_MAX30208_FIFO_STAT_CLR
+#define MAX30208_FIFO_STAT_CLR .fifo_stat_clr = CONFIG_MAX30208_FIFO_STAT_CLR,
+#else
+#define MAX30208_FIFO_STAT_CLR .fifo_stat_clr = 0,
+#endif
```
Kconfig options can be used directly without storing them in config. Why not make these DT properties?, so they can be set per-instance?

trembel replied:
would you make all of my Kconfig options to DT properties?

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,346 @@
+	\
+	static struct max30208_data max30208_data_##n = {	\
+		.raw_buffer = {	\
+			.size = CONFIG_MAX30208_RINGBUFFER_SIZE * MAX30208_BYTES_PER_VAL,	\
```
`ARRAY_SIZE`?

gmarull commented on drivers/sensor/max30208/max30208_trigger.c:
```diff
@@ -0,0 +1,246 @@
+
+	gpio_flags_t flags = enable ? GPIO_INT_EDGE_TO_ACTIVE : GPIO_INT_DISABLE;
+
+	gpio_pin_interrupt_configure_dt(&config->gpio_int, flags);
```
this returns error codes, should be checked.

gmarull commented on drivers/sensor/max30208/max30208_trigger.c:
```diff
@@ -0,0 +1,246 @@
+	int ret = gpio_pin_configure_dt(&config->gpio_int, GPIO_INPUT);
+
+	if (ret < 0) {
+		LOG_DBG("Failed to set gpio as input!");
```
`LOG_ERR`

trembel commented:
Open Questions to me:
> `SENSOR_CHAN_ALL` should also be supported, see https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/sensor.h#L513

As written, should this also be for `channel_get`? IMO not.

> is it worth reading just to log? what if logging is disabled? Isn't it sufficient to warn about buffer being too small?

since one item must be poped on a full ringbuffer, I could just pop and not store it, but I don't see a lot of advantage here..

> why using a struct to read a single byte?

Its a bitfield struct. As it implies the same runtime cost as using a raw byte, I think its better here as specific fields can be accessed by a meaningful name.

carlescufi commented:
@gmarull can you please take another look?

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,329 @@
+		return -ENODATA;
+	}
+
+	int ret = ring_buf_get(&data->raw_buffer, local_buffer, MAX30208_BYTES_PER_VAL);
```
this function returns `uint32_t`, so error handling here is incorrect. ref https://docs.zephyrproject.org/latest/reference/data_structures/ring_buffers.html#c.ring_buf_get

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,329 @@
+	struct max30208_data *data = dev->data;
+
+	uint8_t fifo_data_counter = 0U;
+	int ret =
+		i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_FIFO_DATA_CTR, &fifo_data_counter);
```
formatting can be improved here:
```
	int ret = i2c_reg_read_byte_dt(&config->bus,
				       MAX30208_REG_FIFO_DATA_CTR,
				       &fifo_data_counter);
```

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,329 @@
+	}
+
+	/* Read data ready */
+	uint8_t local_buffer[MAX30208_BYTES_PER_VAL * MAX30208_FIFO_SIZE];
```
define at the top of the function (per convention)

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,329 @@
+		if (ret < 0) {
+			return ret;
+		}
+		if (i == MAX30208_POLL_TRIES) {
```
you'll be running `MAX30208_POLL_TRIES + 1` this way

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,329 @@
+	if (ret < 0) {
+		LOG_ERR("Could not get Part ID");
+		return ret;
+	}
+	if (part_id != MAX30208_PART_ID) {
```
nit: blank line here

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,329 @@
+#endif
+
+#define MAX30208_INIT(n)	\
+	static struct max30208_config max30208_config_##n = {	\
+	    .bus = I2C_DT_SPEC_INST_GET(n),	\
+	    .fifo = {	\
+			.fifo_a_full = DT_INST_PROP(n, fifo_a_full),	\
+			.config2 = {	\
+				.fifo_ro = DT_INST_PROP(n, fifo_rollover_en),	\
+				.a_full_type = DT_INST_PROP(n, fifo_a_full_type),	\
+				.fifo_stat_clr = DT_INST_PROP(n, fifo_stat_clr),	\
+				.flush_fifo = 0,	\
+			},	\
+		},	\
+	    .gpio_setup = {	\
+		.gpio0_mode = DT_INST_PROP(n, gpio0_mode),	\
+		.gpio1_mode = DT_INST_PROP(n, gpio1_mode),	\
+	    },	\
+		MAX30208_GPIO_TRIGGER(n)	\
+	};	\
+	\
+	static struct max30208_data max30208_data_##n = {	\
+		.raw_buffer = {	\
+			.size = MAX30208_ARRAY_SIZE,	\
+			.buf = { .buf8 = max30208_data_##n._ring_buffer_data_raw_buffer }	\
+		}	\
+	};	\
+	\
```
nit: push all backslashes to the same column

mbolivar-nordic commented on dts/bindings/sensor/maxim,max30208.yaml:
```diff
@@ -0,0 +1,76 @@
+# Copyright (c) 2021, Silvano Cortesi
```
This binding needs to justify its use of defaults; see https://docs.zephyrproject.org/latest/guides/dts/bindings.html#general-rules for details.

mbolivar-nordic commented:
+1 for the binding. I didn't review the rest.

trembel commented:
what is blocking this one?

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,323 @@
+	const struct max30208_config *config = dev->config;
+
+	/* Trigger one measurement */
+	int ret = i2c_reg_write_byte_dt(&config->bus, MAX30208_REG_TEMP_SETUP,
+					MAX30208_CONVERT_T_MASK);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
```
nit: just return here

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,323 @@
+
+	uint32_t ret = ring_buf_get(&data->raw_buffer, local_buffer, MAX30208_BYTES_PER_VAL);
+
+	if (ret == 0) {
+		LOG_ERR("Couldn't get data from ringbuffer");
+	}
+
+	/* Concat two bytes to a int16 value */
+	int16_t temp_val = (int16_t)sys_get_be16(local_buffer);
+
+	/* Convert raw value into temperature */
+	val->val1 = temp_val / MAX30208_ONE_DEGREE;
+	val->val2 = (temp_val % MAX30208_ONE_DEGREE) * MAX30208_LSB_e6;
+
+	return 0;
+}
```
this will succeed even if there is no data

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,323 @@
+	struct max30208_status status;
+
+	/* Read data ready */
+	int ret = i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_INT_STS,
+					MAX30208_STC_TO_P(status));
```
nit: align second line to character after `(`

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,323 @@
+	uint8_t local_buffer[MAX30208_BYTES_PER_VAL * MAX30208_FIFO_SIZE];
+
+	/* Read data ready */
+	int ret = i2c_reg_read_byte_dt(&config->bus,
+					MAX30208_REG_FIFO_DATA_CTR,
+					&fifo_data_counter);
```
nit: align to `(`

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,323 @@
+	}
+
+	/* Check the part id to make sure this is MAX30208 */
+	int ret = i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_PART_ID, &part_id);
+
+	if (ret < 0) {
```
this, an others: blank line not needed (please be consistent)

gmarull commented on drivers/sensor/max30208/max30208.c:
```diff
@@ -0,0 +1,323 @@
+
+#define MAX30208_INIT(n)									\
+	static struct max30208_config max30208_config_##n = {					\
+	    .bus = I2C_DT_SPEC_INST_GET(n),							\
+	    .fifo = {										\
+			.fifo_a_full = DT_INST_PROP(n, fifo_a_full),				\
+			.config2 = {								\
+				.fifo_ro = DT_INST_PROP(n, fifo_rollover_en),			\
+				.a_full_type = DT_INST_PROP(n, fifo_a_full_type),		\
+				.fifo_stat_clr = DT_INST_PROP(n, fifo_stat_clr),		\
+				.flush_fifo = 0,						\
+			},									\
+		},										\
+	    .gpio_setup = {									\
+		.gpio0_mode = DT_INST_PROP(n, gpio0_mode),					\
+		.gpio1_mode = DT_INST_PROP(n, gpio1_mode),					\
+	    },											\
+		MAX30208_GPIO_TRIGGER(n)							\
+	};											\
```
nit: mixes spaces/tabs

gmarull reviewed (commented):
Some minor comments, LGTM otherwise

gmarull commented on samples/sensor/max30208/sample.yaml:
```diff
@@ -0,0 +1,11 @@
+sample:
+  description: TBD
+  name: TBD
```
needs update

gmarull commented on samples/sensor/max30208/boards/nrf52dk_nrf52832.overlay:
```diff
@@ -0,0 +1,17 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&i2c0 {
+    max30208@50 {
+        compatible = "maxim,max30208";
+        reg = <0x50>;
+        label = "MAX30208";
+        int-gpios = <&gpio0 22 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
+	fifo-a-full = <2>;
+/*      gpio0-mode = <3>; */
+	status = "okay";
+    };
+};
```
mixes tabs and spaces, why commenting `gpio0-mode`?

gmarull commented on samples/sensor/max30208/boards/nrf52dk_nrf52832.overlay:
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2019, Linaro Ltd.
```
`Linaro`?

gmarull commented on samples/sensor/max30208/sample.yaml:
```diff
@@ -0,0 +1,11 @@
+  sample.sensor.max30208:
+    harness: sensor
+    tags: sensors
+    platform_allow: nrf52dk_nrf52832
```
This sample seems to be generic enough, so it can run on more boards. Better use `filter: dt_compat_enabled("maxim,max30208")`. See for example `samples/sensor/lps22hh/sample.yaml`.

gmarull commented on drivers/sensor/max30208/max30208_trigger.c:
```diff
@@ -0,0 +1,263 @@
+	temp += (MAX30208_ONE_DEGREE * val->val2) / 1000;
+
+	/* Write the alarm configuration register */
+	int ret = i2c_reg_write_byte_dt(&config->bus, reg_addr[1], (uint8_t)((temp >> 8) & 0x00FF));
+
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize Alarm MSB!");
+		return ret;
+	}
+	ret = i2c_reg_write_byte_dt(&config->bus, reg_addr[0], (uint8_t)(temp & 0x00FF));
+	if (ret < 0) {
```
there seems to be inconsistencies still, please revisit

```c
	int ret;

	ret = i2c_reg_write_byte_dt(&config->bus, reg_addr[1], (uint8_t)((temp >> 8) & 0x00FF));
	if (ret < 0) {
		LOG_ERR("Failed to initialize Alarm MSB!");
		return ret;
	}

	ret = i2c_reg_write_byte_dt(&config->bus, reg_addr[0], (uint8_t)(temp & 0x00FF));
	if (ret < 0) {
```

trembel committed:
drivers: sensor: max30208: Added sample for MAX30208

Added sample for MAX30208 Maxim Integrated I2C temperature
sensor. Tested on nrf52dk.

Signed-off-by: Silvano Cortesi <silvano.cortesi@hotmail.com>

mbolivar-nordic reviewed (dismissed):
for the DT portions

MaureenHelm reviewed (changes requested):
Sorry for the late review. Overall this looks very good. It needs a rebase and to add the new sensor to `tests/drivers/build_all/sensor`
